examples/forward_backward.py

```diff
@@ -0,0 +1,194 @@
+    return Z
+
+
+def forward_backward_algorithm(
+    sum_op: AssociativeOp,
+    prod_op: AssociativeOp,
+    factors: List[Funsor],
+    step: Dict[str, str],
+) -> List[Tensor]:
+    """
+    Calculate marginal probabilities:
+    p(x[t], x[t-1] | Y)
+    """
+    step = OrderedDict(sorted(step.items()))
+    drop = tuple("_drop_{}".format(i) for i in range(len(step)))
+    prev_to_drop = dict(zip(step.keys(), drop))
+    curr_to_drop = dict(zip(step.values(), drop))
+    reduce_vars = frozenset(drop)
+
+    # Backward algorithm
+    # beta[t] = p(y[t+1:T]|x[t])
+    # beta[T] = 1
+    data = ops.full_like(factors[0].data, UNITS[prod_op])
+    beta = Tensor(data, factors[0](x_curr="x_prev").inputs, factors[0].dtype)
+    betas = [beta]
+    # inductive steps
+    # beta[t] = trans[t+1] @ beta[t+1]
+    for trans in factors[:0:-1]:
+        beta = prod_op(trans(**curr_to_drop), beta(**prev_to_drop)).reduce(
+            sum_op, reduce_vars
+        )
+        betas.append(beta)
+    else:
+        init = factors[0]
+        Z = prod_op(init(**curr_to_drop), beta(**prev_to_drop)).reduce(
+            sum_op, reduce_vars
+        )
+    betas.reverse()
+
+    # Forward algorithm
+    # p(y[0], x[0])
+    alpha = factors[0]
+    # p(x[0] | Y)
+    marginal = ops.BINARY_INVERSES[prod_op](
+        prod_op(alpha, betas[0](**{"x_prev": "x_curr"})), Z
+    )
+    marginals = [marginal]
+    # inductive steps
+    for trans, beta in zip(factors[1:], betas[1:]):
+        # p(y[0:t], x[t-1], x[t])
+        alpha_trans = prod_op(alpha(**{"x_curr": "x_prev"}), trans)
+        # p(y[0:t], x[t])
+        alpha = alpha_trans.reduce(sum_op, frozenset({"x_prev"}))
+        # p(x[t-1], x[t] | Y)
+        marginal = ops.BINARY_INVERSES[prod_op](
+            prod_op(alpha_trans, beta(**{"x_prev": "x_curr"})), Z
+        )
+        marginals.append(marginal)
+
+    return marginals
+
+
+def main(args):
+    """
+    Compute marginal probabilities p(x[t], x[t-1] | Y) for an HMM:
+
+    x[0] -> ... -> x[t-1] -> x[t] -> ... -> x[T]
+     |              |         |             |
+     v              v         v             v
+    y[0]           y[t-1]    y[t]           y[T]
+
+    Z = p(y[0:T])
+    alpha[t] = p(y[0:t], x[t])
+    beta[t] = p(y[t+1:T] | x[t])
+    trans[t] = p(y[t], x[t] | x[t-1])
+
+    p(x[t], x[t-1] | Y) = alpha[t-1] * beta[t] * trans[t] / Z
+
+    d Z / d trans[t] = alpha[t-1] * beta[t]
+
+    **References:**
+
+    [1] Jason Eisner (2016)
+        "Inside-Outside and Forward-Backward Algorithms Are Just Backprop
+        (Tutorial Paper)"
+        https://www.cs.jhu.edu/~jason/papers/eisner.spnlp16.pdf
+    [2] Zhifei Li and Jason Eisner (2009)
+        "First- and Second-Order Expectation Semirings
+        with Applications to Minimum-Risk Training on Translation Forests"
+        http://www.cs.jhu.edu/~zfli/pubs/semiring_translation_zhifei_emnlp09.pdf
+    """
+    funsor.set_backend("torch")
```

Comment:
Is this necessary?

Reply:
Good catch. Seems to work with numpy too.